_posts/2017-07-05-puppet-ruby-facts.md

```diff
@@ -20,31 +20,31 @@ tags:
+> Facter is Puppet’s cross-platform system profiling library. It discovers and reports per-node facts, which are available in your Puppet manifests as variables.
 
-There are some [core facts](https://docs.puppet.com/facter/latest/core_facts.html) which are processed on all operating system but, two additional types of facts can be used to extend facter; External Facts and Custom Facts
+There are some [core facts](https://docs.puppet.com/facter/latest/core_facts.html) which are processed on all operating systems, but two additional types of facts can be used to extend facter; External Facts and Custom Facts.
 
 **External facts**
 
-[External](https://docs.puppet.com/facter/latest/custom_facts.html#external-facts) facts provide a way to use arbitrary executables or scripts as facts, or set facts statically with structured data. If you’ve ever wanted to write a custom fact in Perl, C, PowerShell, or a one-line text file, this is how.
+[External](https://docs.puppet.com/facter/latest/custom_facts.html#external-facts) facts provide a way to use arbitrary executables or scripts to generate facts as basic key / value pairs. If you’ve ever wanted to write a custom fact in Perl, C or PowerShell, this is how. Alternatively, external facts may contain static structured data in a JSON or YAML file.
 
 **Custom Facts**
 
-[Custom](https://docs.puppet.com/facter/latest/fact_overview.html) facts are written in ruby and have more advanced features and confinements.
+[Custom](https://docs.puppet.com/facter/latest/fact_overview.html) facts are written in Ruby and have more advanced features, allowing for programmatic confinement to specific environments.
 
-Most people when starting out will write PowerShell (or heaven forbid, batch files) external facts which is a good start.  However there are some downsides.  Firstly the execution time PowerShell scripts can be a little slow.  This is mainly due to PowerShell itself starting.  The other downside is you can not tell Unix based operating systems to not try an execute these Windows specific external facts.  While there are some tricks to help stop this, it's far too easy for Windows scripts to log warnings and errors making it harder to figure out when real issues occur.
+Most people new to Facter will write PowerShell scripts as external facts.  However, there is a downside.  The execution time for PowerShell scripts can be a little slow as a result of the time required to start a new PowerShell process for each fact.  Another downside is that Windows will use file extensions to determine if a fact may be executed, while Unix based operating systems will look for the executable bit - it can be easy to forget these rules, especially when building cross-platform modules.  While there are some tricks to help stop this, it's easy for Windows scripts to log warnings and errors making it harder to figure out when real issues occur.
```

Comment:
Removed the bit about execution because on Windows its a static list of extensions, and on non-Windows its `+x` in the filesystem -
 the code lives in https://github.com/puppetlabs/facter/blob/master/lib/src/facts/external/execution_resolver.cc

There were some earlier Puppet versions that spit out poor error messages which made it look like it was trying to execute `ps1` files on Linux when that wasn't really the case (unless they were stored with `+x`)